### PR TITLE
An API for dynamically loading fonts

### PR DESCRIPTION
**Pull request by jason-simmons**
See https://github.com/flutter/flutter/issues/17910

### CODE REVIEW

File: packages/flutter/lib/src/services/font_loader.dart [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+  ///
+  /// The family groups a series of related font assets, each of which defines
+  /// how to render a specific [FontWeight] and [FontStyle] within the family.
+  final String family;
+
+  /// Registers a font asset to be loaded by this font loader.
+  ///
+  /// The [bytes] argument specifies the actual font asset bytes. Currently,
+  /// only TrueType (TTF) fonts are supported.
+  void addFont(Future<ByteData> bytes) {
```
Comment (Hixie) [inline packages/flutter/lib/src/services/font_loader.dart]: maybe add:

```
///
/// It is the responsibility of the caller to handle errors on the provided Future.
```

Comment (Hixie) [reply packages/flutter/lib/src/services/font_loader.dart]: (or will it complete but return failure? we should figure out exactly what this code does if one of these futures fails, and document it)

File: packages/flutter/lib/src/services/font_loader.dart [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    ));
+  }
+
+  /// Loads this font loader's font [family] and all of its associated assets
+  /// into the Flutter engine, making the font available to the current
+  /// application.
+  ///
+  /// This method should only be called once per font loader. Attempts to
+  /// load fonts from the same loader more than once will cause a [StateError]
+  /// to be thrown.
```
Comment (Hixie) [inline packages/flutter/lib/src/services/font_loader.dart]: document how this method works if any of the futures fail here, too

**Comment by Hixie** [comment]
LGTM

File: packages/flutter/lib/src/services/font_loader.dart (hunk 2 above)
Comment (jason-simmons) [reply packages/flutter/lib/src/services/font_loader.dart]: The future returned by `load` will complete with an error if any of the underlying futures completes with an error.  Updating the docs.

File: packages/flutter/lib/src/services/font_loader.dart [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+  /// The family groups a series of related font assets, each of which defines
+  /// how to render a specific [FontWeight] and [FontStyle] within the family.
+  final String family;
+
+  /// Registers a font asset to be loaded by this font loader.
+  ///
+  /// The [bytes] argument specifies the actual font asset bytes. Currently,
+  /// only TrueType (TTF) fonts are supported.
+  void addFont(Future<ByteData> bytes) {
+    _fontFutures.add(bytes.then(
```
Comment (tvolkert) [inline packages/flutter/lib/src/services/font_loader.dart]: Throw if the loader is already loaded.

File: packages/flutter/lib/src/services/font_loader.dart [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+  }
+
+  @protected
+  @visibleForTesting
+  Future<void> loadFont(Uint8List list, String family) {
+    return loadFontFromList(list, fontFamily: family);
+  }
+
+  bool _loaded;
+  List<Future<Uint8List>> _fontFutures;
```
Comment (tvolkert) [inline packages/flutter/lib/src/services/font_loader.dart]: This can be `final`

**Review by tvolkert — APPROVED** [review]
LGTM

**Pull request merged**